lib/active_merchant/billing/gateways/hi_pay.rb

```diff
@@ -0,0 +1,203 @@
+      self.default_currency = 'EUR'
+      self.money_format = :cents
+      self.supported_cardtypes = %i[visa master american_express]
+
+      self.homepage_url = 'https://hipay.com/'
+      self.display_name = 'HiPay'
+
+      def initialize(options = {})
+        requires!(options, :username, :password)
+        @username = options[:username]
+        @password = options[:password]
+        super
+      end
+
+      def purchase(money, payment_method, options = {})
+        authorize(money, payment_method, options.merge({ operation: 'Sale' }))
+      end
+
+      def authorize(money, payment_method, options = {})
+        MultiResponse.run do |r|
+          if payment_method.is_a?(CreditCard)
+            response = r.process { tokenize(payment_method, options) }
+            card_token = response.params['token']
+          elsif payment_method.is_a?(String)
+            card_token, payment_product = payment_method.split('|')
+          end
+
+          post = {}
+          post[:payment_product] = payment_product&.downcase || payment_method.brand
+
+          post[:operation] = options[:operation] || 'Authorization'
+          add_product_data(post, options)
+          add_invoice(post, money, options)
+          post[:cardtoken] = card_token
+          r.process { commit('order', post) }
+        end
+      end
+
+      def capture(money, authorization, options)
+        post = {}
+        post[:operation] = 'capture'
+        post[:currency] = (options[:currency] || currency(money))
+
+        commit('capture', post, { transaction_reference: authorization })
+      end
+
+      def store(payment_method, options = {})
+        tokenize(payment_method, options.merge({ multiuse: '1' }))
+      end
+
+      def scrub(transcrip)
+        # code
+      end
+
+      private
+
+      def add_product_data(post, options)
+        post[:orderid] = options[:order_id] if options[:order_id]
+        post[:description] = options[:description]
+
+        # optional
+        # basket=> refer https://support.hipay.com/hc/fr/articles/115001660469-Payment-Gateway-Shopping-cart-management
+        post[:basket] = options[:basket] if options[:basket]
+        post[:long_description] = options[:long_description] if options[:long_description]
+        post[:shipping] = options[:shipping_fee] if options[:shipping_fee]
+        post[:tax] = options[:tax] if options[:tax]
+        post[:tax_rate] = options[:tax_rate] if options[:tax_rate]
+      end
+
+      def add_invoice(post, money, options)
+        post[:currency] = (options[:currency] || currency(money))
+        post[:amount] = money
+      end
+
+      def add_credit_card(post, credit_card)
+        post[:card_number] = credit_card.number
+        post[:card_expiry_month] = credit_card.month
+        post[:card_expiry_year] = credit_card.year
+        post[:card_holder] = credit_card.name
+        post[:cvc] = credit_card.verification_value
+      end
+
+      def add_address(post, options)
```

Comment:
🔴  Warning 🔴 

The `billing_address` data is not being included in any request.

Reply:
Ups, I missed.